src/iop/rawdenoise.c

```diff
@@ -33,30 +34,81 @@
 #include <stdlib.h>
 #include <strings.h>
 
-DT_MODULE_INTROSPECTION(1, dt_iop_rawdenoise_params_t)
+DT_MODULE_INTROSPECTION(2, dt_iop_rawdenoise_params_t)
+
+#define DT_IOP_RAWDENOISE_INSET DT_PIXEL_APPLY_DPI(5)
+#define DT_IOP_RAWDENOISE_RES 64
+#define DT_IOP_RAWDENOISE_BANDS 5
+
+typedef enum dt_iop_rawdenoise_channel_t {
+  rawdenoise_all = 0,
+  rawdenoise_R = 1,
+  rawdenoise_G = 2,
+  rawdenoise_B = 3,
+  rawdenoise_none = 4
+} dt_iop_rawdenoise_channel_t;
 
 typedef struct dt_iop_rawdenoise_params_t
 {
   float threshold;
+  float x[rawdenoise_none][DT_IOP_RAWDENOISE_BANDS], y[rawdenoise_none][DT_IOP_RAWDENOISE_BANDS];
 } dt_iop_rawdenoise_params_t;
 
 typedef struct dt_iop_rawdenoise_gui_data_t
 {
   GtkWidget *stack;
+  dt_draw_curve_t *transition_curve; // curve for gui to draw
+
   GtkWidget *box_raw;
   GtkWidget *threshold;
   GtkWidget *label_non_raw;
+  GtkDrawingArea *area;
+  GtkNotebook *channel_tabs;
+  double mouse_x, mouse_y, mouse_pick;
+  float mouse_radius;
+  dt_iop_rawdenoise_params_t drag_params;
+  int dragging;
+  int x_move;
+  dt_iop_rawdenoise_channel_t channel;
+  float draw_xs[DT_IOP_RAWDENOISE_RES], draw_ys[DT_IOP_RAWDENOISE_RES];
+  float draw_min_xs[DT_IOP_RAWDENOISE_RES], draw_min_ys[DT_IOP_RAWDENOISE_RES];
+  float draw_max_xs[DT_IOP_RAWDENOISE_RES], draw_max_ys[DT_IOP_RAWDENOISE_RES];
 } dt_iop_rawdenoise_gui_data_t;
 
 typedef struct dt_iop_rawdenoise_data_t
 {
   float threshold;
+  dt_draw_curve_t *curve[rawdenoise_none];
+  dt_iop_rawdenoise_channel_t channel;
+  float force[rawdenoise_none][DT_IOP_RAWDENOISE_BANDS];
 } dt_iop_rawdenoise_data_t;
 
 typedef struct dt_iop_rawdenoise_global_data_t
 {
 } dt_iop_rawdenoise_global_data_t;
 
+int legacy_params(dt_iop_module_t *self, const void *const old_params, const int old_version, void *new_params,
+                  const int new_version)
+{
+  if(old_version == 1 && new_version == 2)
+  {
+    dt_iop_rawdenoise_params_t *o = (dt_iop_rawdenoise_params_t *)old_params;
```

Comment:
I added a comment to explain how it works.
I used this as I saw this was the way it was done in denoiseprofile and nlmeans for example.

But if you prefer, I can define a new struct so that the changes are more explicit.

Reply:
No, fine I read as if you were copying the would struct with memcpy() and that you were reading past the old struct. I don't this that now! Good to me, you are copying the field one by one.